### PR TITLE
Allow nodes to return from being missing.

### LINKED ISSUE #684: "No nodes with metrics" while etcd shows my local node has metrics from 14 seconds ago
```
mikal@marvin:~$ etcdctl get --prefix=True /sf/metrics
/sf/metrics/marvin/
{
    "fqdn": "marvin",
    "metrics": {
        "cpu_available": 8,
        "cpu_load_1": 0.51,
        "cpu_load_15": 0.75,
        "cpu_load_5": 0.52,
        "cpu_max": 8,
        "cpu_max_per_instance": 16,
        "cpu_total_instance_cpu_time": 0,
        "cpu_total_instance_vcpus": 0,
        "disk_free": 321265045504,
        "disk_read_bytes": 595637563392,
        "disk_total": 433770635264,
        "disk_used": 112505589760,
        "disk_write_bytes": 627252290048,
        "instances_active": 0,
        "instances_total": 0,
        "memory_available": 6464,
        "memory_available_libvirt": 2381,
        "memory_ksm_full_scans": 0,
        "memory_ksm_max_page_sharing": 256,
        "memory_ksm_merge_across_nodes": 0,
        "memory_ksm_pages_shared": 0,
        "memory_ksm_pages_sharing": 0,
        "memory_ksm_pages_to_scan": 0,
        "memory_ksm_pages_unshared": 0,
        "memory_ksm_pages_volatile": 0,
        "memory_ksm_run": 0,
        "memory_ksm_sleep_millisecs": 200,
        "memory_ksm_stable_node_chains": 0,
        "memory_ksm_stable_node_chains_prune_millisecs": 2000,
        "memory_ksm_stable_node_dups": 0,
        "memory_ksm_use_zero_pages": 0,
        "memory_max": 15705,
        "memory_max_libvirt": 15705,
        "memory_total_instance_actual": 0,
        "memory_total_instance_max": 0,
        "network_queue_processing": 0,
        "network_queue_waiting": 0,
        "network_read_bytes": 7083703026,
        "network_write_bytes": 4873745523,
        "node_queue_processing": 0,
        "node_queue_waiting": 0
    },
    "timestamp": 1612673015.5577867
}
mikal@marvin:~$ python3
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import time
>>> time.time()
1612673024.8014984
>>> print(time.time() - 1612673015.5577867)
14.693849563598633
```

### PR DESCRIPTION
**Pull request by mikalstill**
This is especially useful for laptop developer environments.
Fixes #684.

This is a commit to master because I think its worth a release as a hotfix. I'll rebase develop when it merges here.

### CODE REVIEW

File: shakenfist/daemons/cleaner.py [hunk 1]
```diff
@@ -181,10 +182,19 @@ def run(self):
                     ni['uuid']).info('Hard deleting network interface')
                 db.hard_delete_network_interface(ni['uuid'])
 
-            # Find nodes which have been offline for a long time
             for n in Nodes([node_inactive_states_filter]):
                 age = time.time() - n.last_seen
-                if age > config.NODE_CHECKIN_MAXIMUM * 10:
+
+                # Find nodes which have returned from being missing
```
Comment (mikalstill) [inline shakenfist/daemons/cleaner.py:188]: To be honest, I am unsure if this is the right way to do this, or just to set the node state to created every time the node is observed. The second way would involve a much higher etcd write load, but this way feels ... clunky.

File: shakenfist/scheduler.py [hunk 2]
```diff
@@ -179,11 +194,17 @@ def _find_most_matching_images(self, requested_images, candidates):
...
+            # Refresh metrics if its too old, or there are no nodes.
             diff = time.time() - self.metrics_updated
-            if diff > config.get('SCHEDULER_CACHE_TIMEOUT'):
+            log_ctx.debug('Metrics are %.02f seconds old, max is %.02f seconds. '
+                          'Cache has %d elements.'
+                          % (diff, config.SCHEDULER_CACHE_TIMEOUT, len(self.metrics)))
+            if diff > config.SCHEDULER_CACHE_TIMEOUT or len(self.metrics) == 0:
                 self.refresh_metrics()
+                log_ctx.debug('Cache has %d elements after refresh.'
+                              % len(self.metrics))
```
Comment (mandoonandy) [inline shakenfist/scheduler.py:207]: For efficiency, log messages should use the args and kwargs of debug() rather sending an evaluated string. The logger delays evaluation until it checks the logging level.

**Review by mandoonandy — CHANGES_REQUESTED** [review]
tiny non functional requests

File: shakenfist/scheduler.py [hunk 3]
```diff
@@ -43,20 +44,34 @@ def get_network_node():
 
 class Scheduler(object):
     def __init__(self):
+        # This UUID doesn't really mean much, except as a way of tracing the
+        # behaviour of a single instance of the scheduler object in logs.
+        self.__uuid = str(uuid.uuid4())
+        self.log = LOG.with_field('scheduler instance', self.__uuid)
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/scheduler.py:50]: Done.

File: shakenfist/scheduler.py (hunk 2 above)
Comment (mikalstill) [reply shakenfist/scheduler.py:207]: Done.

**Review by mandoonandy — APPROVED** [review]
LGTM

**Pull request merged**